postgis/lwgeom_in_marc21.c

```diff
@@ -0,0 +1,527 @@
+ *
+ * Copyright (C) 2021 University of Münster (WWU), Germany
+ * Written by Jim Jones <jim.jones@uni-muenster.de>
+ *
+ **********************************************************************/
+
+PG_FUNCTION_INFO_V1(ST_GeomFromMARC21);
+Datum ST_GeomFromMARC21(PG_FUNCTION_ARGS) {
+	GSERIALIZED *geom;
+	LWGEOM *lwgeom;
+	xmlDocPtr xmldoc;
+	text *xml_input;
+	int xml_size;
+	char *xml;
+	xmlNodePtr xmlroot = NULL;
+
+	if (PG_ARGISNULL(0)) PG_RETURN_NULL();
+
+	xml_input = PG_GETARG_TEXT_P(0);
+	xml = text_to_cstring(xml_input);
+	xml_size = VARSIZE_ANY_EXHDR(xml_input);
+
+	xmlInitParser();
+	xmldoc = xmlReadMemory(xml, xml_size, NULL, NULL, XML_PARSE_SAX1);
+
+	if (!xmldoc || (xmlroot = xmlDocGetRootElement(xmldoc)) == NULL) {
+		xmlFreeDoc(xmldoc);
+		xmlCleanupParser();
+		lwpgerror("invalid MARC21/XML document.");
+	}
+
+	lwgeom = parse_marc21(xmlroot);
+
+	xmlFreeDoc(xmldoc);
+	xmlCleanupParser();
+
+	if (lwgeom == NULL) {
+
+		lwgeom_free(lwgeom);
+		PG_RETURN_NULL();
+
+	}
+
+	geom = geometry_serialize(lwgeom);
+
+	lwgeom_free(lwgeom);
+
+	PG_RETURN_POINTER(geom);
+}
+
+static int is_literal_valid(char *literal) {
+
+	int j;
+	int num_dec_sep;
+	int coord_start;
+	int literal_length = strlen(literal);
+
+	POSTGIS_DEBUGF(2, "is_literal_valid called (%s)", literal);
+
+	if (literal_length < 3)	return LW_FALSE;
+
+	coord_start = 0;
+	num_dec_sep = 0;
+
+	/**
+	 * Shortest allowed literal:
+	 *   containing cardinal direction >  cardinal direction + degree with 3 digits
+	 *   no cardinal direction >  degree with 3 digits
+	 *
+	 * The variable "coord_start" stores the position where the coordinates start (0 or 1),
+	 * so that the script can identify where each literal element begins to validate its content.
+	 */
+
+	if (literal[0] == 'N' || literal[0] == 'E' || literal[0] == 'S'	|| literal[0] == 'W' || literal[0] == '+' || literal[0] == '-') {
+
+		if (literal_length < 4) {
+			POSTGIS_DEBUGF(3, "  invalid literal length (%d): \"%s\"", literal_length, literal);
+			return LW_FALSE;
+		}
+
+		coord_start = 1;
+	}
+
+	for (j = coord_start; j < literal_length; j++) {
+
+		if (!isdigit(literal[j])) {
+
+
+			if (j < 3) {
+
+				/**
+				 * The first three characters represent the degrees and they must be numeric.
+				 */
+
+				POSTGIS_DEBUGF(3,"  invalid character '%c' at the degrees section: \"%s\"",	literal[j], literal);
+				return LW_FALSE;
+
+			}
+
+			/**
+			 * Decimal degrees are encoded with after the 3rd character with either a dot or a comma.
+			 * Only one decimal separator is allowed.
+			 */
+			if (literal[j] == '.' || literal[j] == ',') {
+
+				num_dec_sep++;
+
+				if (num_dec_sep > 1) return LW_FALSE;
+
+			} else {
+				POSTGIS_DEBUGF(3, "  invalid character '%c' in %d: \"%s\"",	literal[j], j, literal);
+				return LW_FALSE;
+
+			}
+
+		}
+
+	}
+
+	POSTGIS_DEBUGF(2, "=> is_literal_valid returns LW_TRUE for \"%s\"",	literal);
+	return LW_TRUE;
+
+}
+
+static double parse_geo_literal(char *literal) {
```

Comment:
What is the rationale for splitting functionality between here and liblwgeom? If you're going to use liblwgeom, then it makes sense to (a) have all the necessarily functionality there (b) that means the calling functions in postgis/ do little more than deserialize, maybe check empty/null, and then call into liblwgeom. And a main reason for putting code into liblwgeom is so you can unit test it in cunit, which allows for things like memory leak checking, etc. Having everything in your ./postgis in/out files wouldn't be the end of the world.

Reply:
Now that you mention it, I also do not see any reason to split the functionalities between `postgis/` and `liblwgeom/`. I most certainly got confused with the kml functions, as the output function is in `liblwgeom/` and the input in `postgis/` - I took the kml functions as a model to know where/how to write my code . Should I move the input and output functions to `postgis/`? Thanks for the thorough evaluation!

Reply:
Yes, unifying them in the one place will make everything simpler and I don't see any huge leverage from being in ./liblwgeom (except being able to valgrind it, which is always handy)